File: Upcoder/Lập trình cơ bản/4.Hàm/Đạo hàm/DAOHAM.py
#Code này chưa được chấm trên upcoder

def xuat_da_thuc(a, n, k):
    da_thuc = []
    for i in range(k):
        if a[i] != 0:  # Không xuất hằng số 0
            if n[i] == 0:
                da_thuc.append(f"{a[i]}")
            elif n[i] == 1:
                if a[i] == 1:
                    da_thuc.append("x")
                else:
                    da_thuc.append(f"{a[i]}x")
            else:
                if a[i] == 1:
                    da_thuc.append(f"x^{n[i]}")
                else:
                    da_thuc.append(f"{a[i]}x^{n[i]}")
    return " + ".join(da_thuc)

def tinh_gia_tri_da_thuc(a, n, k, x):
    S = 0
    for i in range(k):
        S += a[i] * (x ** n[i])
    return S

def dao_ham_da_thuc(a, n, k):
    da_ham = []
    for i in range(k):
        if n[i] > 0 and a[i] != 0:  # Chỉ tính đạo hàm nếu bậc lớn hơn 0
            he_so_dao_ham = a[i] * n[i]
            bac_dao_ham = n[i] - 1

            if bac_dao_ham == 0:
                da_ham.append(f"{he_so_dao_ham}")
            elif bac_dao_ham == 1:
                da_ham.append(f"{he_so_dao_ham}x")
            else:
                da_ham.append(f"{he_so_dao_ham}x^{bac_dao_ham}")

    if not da_ham:  # Nếu không có đơn thức nào
        return "0"
    return " + ".join(da_ham)

def main():
    k = int(input())
    a = list(map(float, input().split()))
    n = list(map(int, input().split()))
    x = float(input())

    da_thuc = xuat_da_thuc(a, n, k)
    gia_tri = tinh_gia_tri_da_thuc(a, n, k, x)
    da_ham = dao_ham_da_thuc(a, n, k)

    print(da_thuc)
    print(gia_tri)
    print(da_ham)

if __name__ == "__main__":
    main()

File: Upcoder/Lập trình cơ bản/4.Hàm/Đạo hàm/DAOHAM.cs
using System;
namespace DaoHamDaThuc
{
    public class Program
    {
        static void XuatDaThuc(double[] a, int[] n, int k)
        {
            bool first = true;
            for (int i = 0; i < k; i++)
            {
                if (a[i] != 0)
                {
                    if (!first) Console.Write(" + ");
                    first = false;

                    if (n[i] == 0)
                        Console.Write(a[i]);
                    else if (n[i] == 1)
                    {
                        if (a[i] == 1) Console.Write("x");
                        else Console.Write(a[i] + "x");
                    }
                    else
                    {
                        if (a[i] == 1) Console.Write("x^" + n[i]);
                        else Console.Write(a[i] + "x^" + n[i]);
                    }
                }
            }
            Console.WriteLine();
        }

        static double TinhGiaTriDaThuc(double[] a, int[] n, int k, double x)
        {
            double S = 0;
            for (int i = 0; i < k; i++)
            {
                S += a[i] * Math.Pow(x, n[i]);
            }
            return S;
        }

        static void DaoHamDaThuc(double[] a, int[] n, int k)
        {
            bool first = true;
            int cnt = 0;
            for (int i = 0; i < k; i++)
                if (n[i] == 0)
                    cnt++;
            
            if (cnt == k)
            {
                Console.WriteLine(0);
                return;
            }

            for (int i = 0; i < k; i++)
            {
                if (n[i] > 0 && a[i] != 0)
                {
                    if (!first) Console.Write(" + ");
                    first = false;

                    double heSoDaoHam = a[i] * n[i];
                    int bacDaoHam = n[i] - 1;

                    if (bacDaoHam == 0)
                        Console.Write(heSoDaoHam);
                    else if (bacDaoHam == 1)
                        Console.Write(heSoDaoHam + "x");
                    else
                        Console.Write(heSoDaoHam + "x^" + bacDaoHam);
                }
            }
        }

        static void Main()
        {
            int k = int.Parse(Console.ReadLine());
            string []input1 = Console.ReadLine().Split(' ');
            string []input2 = Console.ReadLine().Split(' ');
            double[] a = new double[k];
            int[] n = new int[k];

            for (int i = 0; i < k; i++)
                a[i] = double.Parse(input1[i]);

            for (int i = 0; i < k; i++)
                n[i] = int.Parse(input2[i]);

            double x = double.Parse(Console.ReadLine());
            if (TinhGiaTriDaThuc(a, n, k, x) == 0)
            {
                Console.WriteLine(0);
                Console.WriteLine(0);
                Console.Write(0);
                return;
            }

            XuatDaThuc(a, n, k);
            double ketQua = TinhGiaTriDaThuc(a, n, k, x);
            Console.WriteLine(ketQua);
            DaoHamDaThuc(a, n, k);
        }
    }
}
